Stop using VersionedObject in resource.Builder, use unstructured

smarterclayton opened the pull request:
Remove the need for kubectl callers to distinguish between unstructured and versioned / type aware builders.  The factory should create a single builder than can be set to return unstructured objects.  Callers can then use one of the new helpers on `resource.Info` to convert the objects into the desired form - `Internal()` for printers, `Typed()` for external versions, and `Unstructured()` to ensure the object is in the right state.  Leverages the new scheme support for unstructured conversion so that higher level callers can perform best effort conversion (get typed versions if you have them, otherwise use default behavior).  

`get.go` demonstrates this by removing the previous logic that depended on the underlying scheme.  Other commands are updated to be consistent.

Includes #55650 and #55647.

smarterclayton committed:
Revert "refactor builder in kubectl factory"

This reverts commit 06c5be98028d6c045ae438498b46262ea1035b9c.

smarterclayton committed:
Unify unstructured and versioned object in resource.Builder

resource.Builder should be aware of both paths, and the caller is
responsible for determining the different path via use.

smarterclayton committed:
Builder should allow moving between unstructured / versioned easily

Demonstrate its use with get.go by removing the need to call decode with
a hardcoded scheme. Add tests to get_test.go to restore proving that
unstructured conversion is possible.

smarterclayton committed:
Remove use of VersionedObject and simplify builder in generic methods

Reduce all uses of Unstructured to the simpler form, and avoid asking
for mapper or typer unless it is required. Use Typed() for places that
previously used VersionedObject, and remove paths for versioned objects
from code that is now using unstructured.

smarterclayton commented:
@kubernetes/sig-cli-api-reviews

zjj2wry commented on pkg/kubectl/cmd/set/set_env.go:
```diff
@@ -309,21 +309,22 @@ func (o *EnvOptions) RunEnv(f cmdutil.Factory) error {
 	}
 	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) ([]byte, error) {
-		_, err := o.UpdatePodSpecForObject(info.VersionedObject, func(spec *v1.PodSpec) error {
+		info.Object = info.AsVersioned()
```
this will ignore convert err. when convert error will return `the object is not a pod or does not have a pod template` for users. i think this error cannot help users...

but i am ok use `info.AsVersioned()` in `PrintObject`

smarterclayton replied:
A convert error can only happen if the object is not a valid type, which means we aren't compiled with it, which means we weren't going to be able to support it here anyway.  We already have types that aren't compiled in, so the helper already has to deal with that.

smarterclayton replied:
I would say that we should be as explicit as we can when something is not in the known type.  We should probably do that up front in the resource.Builder (have a RequireVersioned() flag) that will accumulate errors for anything that isn't versioned.

However, when doing anything from disk we have to be pretty tolerate of things that aren't versioned (i.e. `kubectl set image -f FILE_THAT_HAS_MIXED_OBJECTS`) - that command shouldn't fail for objects it doesn't recognize.  So tolerating untyped objects in set commands is going to be pretty important.

smarterclayton committed:
Remove unstructured error checking from affected code

Also remove error messages that depended on ObjectKinds() - future
changes will potentially remove this interface and the replacements here
are equivalent.

smarterclayton committed:
Apply set last applied should not need to fetch latest

Latest always refreshes from the server, and the server will return the
updated object. In this case, we want to only access the local file.

smarterclayton committed:
Add a lazy discovery interface for Unstructured

Delays the error until the first call and then preserves it for others.
More closely matches the intent of the Object() calls. Loaders are now
lazy and don't need to return errors directly.

Sets the stage for collapsing unstructured and structured builders
together.

smarterclayton committed:
Unify Object and UnstructuredObject

The unified RESTMapper and Typer follow the new rules, but on error will
fallback to the legacy path (while still supporting Unstructured
objects). This allows callers to handle the appropriate distinction
themselves if necessary.

Add a LocalParam() method to the resource.Builder that DRYs up a large
chunk of complicated code in set commands.

smarterclayton commented:
@deads2k UnstructuredObject() is now dead.  Can you look at the last two commits and give your blessing / opinion?

smarterclayton committed:
All commands must declare Unstructured or Internal

Callers must take a dependency on one or the other set of conversions
and default client behavior. Future changes may add a Versioned() type,
but this is an accurate reflection of current code state.

smarterclayton commented:
Last commit added that forces all resource.Builder users to decide on Unstructured or Internal.  I'll resist the urge to keep refactoring at this point - this is enough to shrink get.go, give us a clear anchor point to remove unstructured, and removes the accrued debt around not being able to rationalize unstructured and versioned types.

deads2k commented on pkg/kubectl/resource/builder.go:
```diff
@@ -166,16 +171,42 @@ func (b *Builder) FilenameParam(enforceNamespace bool, filenameOptions *Filename
+// into the internal form if necessary. Using internal types is lossy - fields added
+// to the server will not be seen by the client code and may result in failure. Only
+// use this mode when working offline, or when generating patches to send to the server.
+// TODO: Introduce a Versioned() form that performs this same function, but biases
```
The warning on this method would remain identical.  For a given snapshot of external types, the internal ones are just as good.

deads2k commented on pkg/kubectl/cmd/annotate.go:
```diff
@@ -189,13 +189,10 @@ func (o AnnotateOptions) RunAnnotate(f cmdutil.Factory, cmd *cobra.Command) erro
-	}
-	b = b.ContinueOnError().
+	b := f.NewBuilder().
+		Unstructured().
```
@juanvallejo  didn't `Unstructured`  attempt to make a remote call and that broke `--local` since it tried to contact the server.

juanvallejo replied:
Yes, I believe `Unstructured()` makes use of the discovery client to retrieve mappings

smarterclayton replied:
Yes, and now unstructured fails back to using the innate discovery.

deads2k commented on pkg/kubectl/cmd/apply_set_last_applied.go:
```diff
@@ -112,7 +112,7 @@ func (o *SetLastAppliedOptions) Complete(f cmdutil.Factory, cmd *cobra.Command)
 
 	var err error
-	o.Mapper, o.Typer = f.UnstructuredObject()
+	o.Mapper, o.Typer = f.Object()
```
I don't have all the code in my head, but this leaps out as moving in the wrong direction.  Why the change?

smarterclayton replied:
There is only Object() now.

smarterclayton replied:
Everything is in the same path.  You ask the builder whether you want Unstructured() or Internal().  You can't opt out.

deads2k commented on pkg/kubectl/cmd/set/set_serviceaccount.go:
```diff
@@ -127,16 +127,16 @@ func (saConfig *serviceAccountConfig) Complete(f cmdutil.Factory, cmd *cobra.Com
-	builder := f.NewBuilder().ContinueOnError().
+	builder := f.NewBuilder().
+		LocalParam(saConfig.local).
+		ContinueOnError().
```
intentional addition?  I like it, but wanted to make sure it was on purpose.

smarterclayton replied:
Yes, was down below.

smarterclayton replied:
I have this TODO somewhere, but local shouldn't be a command specific flag.  It should be a global and enforced at multiple levels.  There are decisions that Factory methods make that should depend on local.

deads2k commented on pkg/kubectl/cmd/util/factory_builder.go:
```diff
@@ -108,12 +103,6 @@ func (f *ring2Factory) PrintObject(cmd *cobra.Command, isLocal bool, mapper meta
 	gvks, _, err := typer.ObjectKinds(obj)
 
-	// fall back to an unstructured object if we get something unregistered
-	if runtime.IsNotRegisteredError(err) {
```
I remember this being important for handling cases where you querying aggregated types and CRDs.

smarterclayton replied:
I now handle unknown types in the fundamental scheme safely for ObjectKinds.

deads2k commented on pkg/kubectl/cmd/util/factory_object_mapping.go:
```diff
@@ -73,33 +73,15 @@ func NewObjectMappingFactory(clientAccessFactory ClientAccessFactory) ObjectMapp
-	discoveryClient, err := f.clientAccessFactory.DiscoveryClient()
-	if err != nil {
-		return nil, nil, err
+		glog.V(3).Infof("Unable to get a discovery client to find server resources, falling back to hardcoded types: %v", err)
```
Why should we continue doing this?

smarterclayton replied:
Fallback for local.  You still need a RESTMapper if you can't contact the server.  If discovery doesn't work, all but a few commands want to fail partially.

deads2k commented on pkg/kubectl/cmd/util/discovery.go:
```diff
@@ -0,0 +1,121 @@
+// NewLazyObjectLoader handles unrecoverable errors when creating a RESTMapper / ObjectTyper by
+// returning those initialization errors when the interface methods are invoked. This defers the
+// initialization and any server calls until a client actually needs to perform the action.
+func NewLazyObjectLoader(fn func() (meta.RESTMapper, runtime.ObjectTyper, error)) (meta.RESTMapper, runtime.ObjectTyper) {
```
Ah, I see.  This is what is letting you remove the special local blocks.

deads2k commented on pkg/kubectl/cmd/util/discovery.go:
```diff
@@ -0,0 +1,121 @@
+func (o *lazyObject) init() error {
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	if o.err != nil || o.loader == nil {
```
separate the if blocks and set the error if the loader is nil.

smarterclayton replied:
No, loader in this case is the indicator that we have already loaded or not.  We clear the loader and then don't continue.

deads2k commented on pkg/kubectl/cmd/util/discovery.go:
```diff
@@ -0,0 +1,121 @@
+	if o.err != nil || o.loader == nil {
+		return o.err
+	}
+	o.mapper, o.typer, o.err = o.loader()
```
check the mapper and typer early and return if they're set.

deads2k commented on pkg/kubectl/cmd/util/discovery.go:
```diff
@@ -0,0 +1,121 @@
+		return o.err
+	}
+	o.mapper, o.typer, o.err = o.loader()
+	o.loader = nil
```
I don't see a need to do this.  mapper and typer shouldn't be nil out of the loader.

smarterclayton replied:
This is the indicator that covers whether loading has happened.

deads2k replied:
> This is the indicator that covers whether loading has happened.

Using it as the trigger opens you up to accidentally leaving nil.  Seems like triggering the other way around avoids that problem.

deads2k commented on pkg/kubectl/cmd/util/discovery.go:
```diff
@@ -0,0 +1,121 @@
+var _ meta.RESTMapper = &lazyObject{}
+var _ runtime.ObjectTyper = &lazyObject{}
+
+func (o *lazyObject) KindFor(resource schema.GroupVersionResource) (schema.GroupVersionKind, error) {
```
I'd like to the deferred RESTMapper moved with the other RESTMappers and not specific to kubectl.

smarterclayton replied:
it's gotta have object typer too, but I can do that.

deads2k commented on pkg/kubectl/cmd/util/factory_builder.go:
```diff
@@ -163,7 +155,7 @@ func (f *ring2Factory) NewBuilder() *resource.Builder {
 
 	unstructuredClientMapperFunc := resource.ClientMapperFunc(f.objectMappingFactory.UnstructuredClientForMapping)
-	unstructuredMapper, unstructuredTyper, _ := f.objectMappingFactory.UnstructuredObject()
+	unstructuredMapper, unstructuredTyper := f.objectMappingFactory.Object()
```
this looks backwards

smarterclayton commented:
/retest

smarterclayton commented:
/retest

mengqiy commented:
bazel test is failing.

mengqiy commented on pkg/kubectl/cmd/util/factory_builder.go:
```diff
@@ -51,17 +50,9 @@ func NewBuilderFactory(clientAccessFactory ClientAccessFactory, objectMappingFac
 func (f *ring2Factory) PrinterForCommand(cmd *cobra.Command, isLocal bool, outputOpts *printers.OutputOptions, options printers.PrintOptions) (printers.ResourcePrinter, error) {
```
> isLocal bool

It seems we don't use it any more. We can drop it.

smarterclayton replied:
Will do that in a follow up - was going to do the following:

1. remove isLocal
2. Collapse OutputOptions into PrintOptions
3. show how that is used by server-side get rendering

juanvallejo replied:
> 2. Collapse OutputOptions into PrintOptions

PR currently open for this: https://github.com/kubernetes/kubernetes/pull/54811

mengqiy commented on pkg/kubectl/resource/visitor.go:
```diff
@@ -81,10 +82,6 @@ type Info struct {
-	// and conversions into its corresponding internal type. This is useful for
-	// reflecting on user intent which may be lost after defaulting and conversions.
-	VersionedObject runtime.Object
 	// Optional, this is the most recent value returned by the server if available
```
Can you please update the comment to make it more helpful?

smarterclayton replied:
Yes

mengqiy reviewed (commented):
I like this PR!

smarterclayton committed:
rollingupdate should use explicit casting

It needs both internal and versioned - use the new helpers to detect
that

smarterclayton commented:
Updated and rebased on the latest changes.  Last thing blocking client side get alpha changes for 1.9

shiywang commented on pkg/kubectl/resource/mapper.go:
```diff
@@ -44,17 +34,27 @@ type Mapper struct {
 // if any of the decoding or client lookup steps fail. Name and namespace will be
 // set into Info if the mapping's MetadataAccessor can retrieve them.
 func (m *Mapper) InfoForData(data []byte, source string) (*Info, error) {
-	versions := &runtime.VersionedObjects{}
```
Is kubectl the only consumer of `VersionedObjects{}` do we also need to remove [here](https://github.com/kubernetes/kubernetes/blame/3df3c580b7a318c8fd166296b61d891f6c0359fa/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L93) and append logic [there](https://github.com/kubernetes/kubernetes/blame/3df3c580b7a318c8fd166296b61d891f6c0359fa/staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go#L135-L138) ?

shiywang replied:
Sorry, i think i forget the compatible problem...please omit

smarterclayton replied:
I will be removing VersionedObjects in a future PR.

deads2k commented:
/lgtm
